library/sha256.c

```diff
@@ -64,7 +65,15 @@
  * Capability detection code comes early, so we can disable
  * MBEDTLS_SHA256_USE_A64_CRYPTO_IF_PRESENT if no detection mechanism found
  */
-#if defined(HWCAP_SHA2)
+#if defined(HAVE_GETAUXVAL)
+#if !defined(HWCAP_SHA2)
+/* The same header that declares getauxval() should provide the HWCAP_xxx
+ * constants to analyze its return value. However, the libc may be too
+ * old to have the constant that we need. So if it's missing, assume that
+ * the value is the same one used by the Linux kernel ABI.
+ */
+#define HWCAP_SHA2 (1 << 6)
```

Comment:
I would suggest we do [what we are doing upstream in an OpenSSL PR](https://github.com/openssl/openssl/pull/20305/files#diff-cae93dac08c5c080327a251a304d0cd3aa807abdc27a17c24fa642751d8b555fR135), and use a different name here, such as `MBEDTLS_HWCAP_SHA2`. Then we don't need the `#if !defined(HWCAP_SHA2)` dance.

These constants are effectively part of the ABI, and can't change without breaking any software that uses them (they need to match between the kernel and userland).

Reply:
agree with @tom-cosgrove-arm

Reply:
The constants don't _need_ to match between the `getauxval` library function and the underlying system call. It's rare, but it has happened before that libc and syscall have diverged on bitmasks.

However, in this case, the bitmasks ultimately come from the hardware. So divergence would be extremely unlikely, and I'm ok with defining our own constants.

Reply:
@yuhaoth Should I rewrite this PR, or do you prefer to handle it as part of https://github.com/Mbed-TLS/mbedtls/pull/7078 ?

Reply:
> However, in this case, the bitmasks ultimately come from the hardware

That's not the case. For example, the `getauxval()` API uses one bit to indicate SHA256 support, and another to indicate SHA512 support, whereas `ID_AA64ISAR0_EL1` uses 4 bits for both (0: none; 01b: SHA256; 10b: SHA256 + SHA512)

> it has happened before that libc and syscall have diverged on bitmasks

Agreed, but once userland is expecting particular values, changing them at the kernel side requires corresponding changes in userland, and therefore can't change without breaking any software that uses them (they need to match, even if they don't have the same values).

Reply:
Ah. Then there is more of a risk that a libc would implement its own bit masks. But I'm ok with fixing that if and when it happens. Still, if `sys/auxv.h` does define the `HWCAP_xxx` constant, I don't see why we wouldn't use it.

Reply:
> @yuhaoth Should I rewrite this PR, or do you prefer to handle it as part of #7078 ?

I prefer handle it as part of #7078.

Reply:
> Ah. Then there is more of a risk that a libc would implement its own bit masks. But I'm ok with fixing that if and when it happens. Still, if `sys/auxv.h` does define the `HWCAP_xxx` constant, I don't see why we wouldn't use it.

For me, that makes our headers too complex and not easily maintain.